File: src/Functional/OCaml/proxy.ml
type backend = { mutable created : int; mutable fetches : int }

let make_proxy backend =
  let cache = Hashtbl.create 4 in
  let subject_created = ref false in
  fun id ->
    match Hashtbl.find_opt cache id with
    | Some value -> value
    | None ->
        if not !subject_created then (
          backend.created <- backend.created + 1;
          subject_created := true);
        backend.fetches <- backend.fetches + 1;
        let value = Printf.sprintf "doc(%d)" id in
        Hashtbl.add cache id value;
        value

let () =
  let backend = { created = 0; fetches = 0 } in
  let get_document = make_proxy backend in
  let first = get_document 42 in
  let second = get_document 42 in
  Printf.printf "backend=%d;fetches=%d;first=%s;second=%s\n"
    backend.created backend.fetches first second

File: src/Functional/Scala/Proxy.scala
import scala.collection.mutable

trait DocumentStore {
  def get(id: Int): String
}

final class RemoteDocumentStore extends DocumentStore {
  var fetchCount: Int = 0
  def get(id: Int): String = {
    fetchCount += 1
    "doc(" + id + ")"
  }
}

final class DocumentStoreProxy extends DocumentStore {
  private var backend: Option[RemoteDocumentStore] = None
  private val cache = mutable.Map.empty[Int, String]

  def get(id: Int): String = cache.getOrElseUpdate(id, {
    val real = backend.getOrElse {
      val created = new RemoteDocumentStore
      backend = Some(created)
      created
    }
    real.get(id)
  })

  def backendCount: Int = if (backend.isDefined) 1 else 0
  def fetchCount: Int = backend.map(_.fetchCount).getOrElse(0)
}

object ProxyExample extends App {
  val store = new DocumentStoreProxy
  val first = store.get(42)
  val second = store.get(42)
  println("backend=" + store.backendCount + ";fetches=" + store.fetchCount + ";first=" + first + ";second=" + second)
}
